### PR TITLE
Critical Hits on Focused Arrow Strike and Gunslinger

### LINKED ISSUE #8200: Critical Hits on Focused Arrow Strike and Gunslingers have incorrect damage
### rAthena Hash

7d9a23d

### Client Date

2021-11-03

### Server Mode

Pre-Renewal

### Result

When Focused Arrow Strike crits, it always deals max damage.
Bullet attack is included in criticals.
Chain Action can do criticals.

### Relevant Log Output

_No response_

### Expected Result

When Focused Arrow Strike crits, there is damage variance.
Bullet attack is not included in criticals (arrow ATK is!).
Chain Action cannot do critical hits.

### How to Reproduce

Get high crit chance on a sniper, use Focused Arrow Strike, check damage.
Get high crit chance on a gunslinger, equip bullet, check damage. Equip blood shell, check damage.
Get Chain Action, check if there are critical double hits.

### Official Information

Test results:

Focused Arrow Strike (Job 69, 99 DEX, Bow) vs. Ant Egg:
// Sharpshooting
DMG - 985-1100
DMG Crit - 1260-1336
// True Sight 10
DMG - 1225-1346
DMG Crit - 1528-1657 

Gunslinger (Job 70, 99 LUK, Crimson Bolt) vs. Verit:
Crit With Bullet - 78
Crit With Blood Shell - 78
Crit With Silver Bullet - 117
vs. Golem
Crit With Bullet - 78
Crit With Blood Shell - 78
Crit With Silver Bullet - 78

Grenade Launcher + Lighting Sphere:
Crit vs. Verit - 253
Crit vs. Golem - 253

Sniper (Job 69, 99 LUK, Bow) vs. Verit:
Crit with Arrow - 91
Crit with Oridecon Arrow - 116
vs. Golem
Crit with Arrow - 81
Crit with Oridecon Arrow - 99

### Modifications that may affect results

_No response_

### PR DESCRIPTION
**Pull request by Playtester**
* **Addressed Issue(s)**: #8200 

* **Server Mode**: Pre-Renewal

**Description of Pull Request**: 

- Focused Arrow Strike no longer deals max damage when it crits (pre-re)
- Focused Arrow Strike no longer considers cards that increase/decrease critical damage (pre-re)
- Criticals with any guns are now calculated like melee criticals, not considering bullet attack (pre-re)
- Chain Action can no longer do critical hits (pre-re)
- Fixes #8200

### CODE REVIEW

File: src/map/battle.cpp [hunk 1]
```diff
@@ -4071,6 +4071,14 @@ static void battle_calc_skill_base_damage(struct Damage* wd, struct block_list *
 						break;
 				}
 			}
+			if (skill_id == SN_SHARPSHOOTING || skill_id == MA_SHARPSHOOTING)
+				i &= ~(1); //Sharpshooting just ignores DEF but does not deal max damage
```
Comment (aleos89) [inline src/map/battle.cpp:4075]: Doesn't have to be done in this PR, but just a reminder for anyone. Might be beneficial to turn these flags into constants so we can easily read them without having to scroll around for hidden comments.

**Review by aleos89 — APPROVED** [review]

File: src/map/battle.cpp (hunk 1 above)
Comment (Playtester) [reply src/map/battle.cpp]: Improved

File: src/map/battle.cpp [hunk 2]
```diff
@@ -4039,41 +4039,46 @@ static void battle_calc_skill_base_damage(struct Damage* wd, struct block_list *
...
 			}
 #else
-			i = (is_attack_critical(wd, src, target, skill_id, skill_lv, false)?1:0)|
-				(is_skill_using_arrow(src, skill_id)?2:0)|
-				(skill_id == HW_MAGICCRASHER?4:0)|
-				(!skill_id && sc && sc->getSCE(SC_CHANGE)?4:0)|
-				(skill_id == MO_EXTREMITYFIST?8:0)|
-				(sc && sc->getSCE(SC_WEAPONPERFECTION)?8:0);
+			// Pre-renewal exclusive flags
+			bflag |= (is_skill_using_arrow(src, skill_id) ? BDMG_ARROW : BDMG_NONE);
```
Comment (aleos89) [inline src/map/battle.cpp:4055]: You could make these into normal if statements to get rid of the short circuit else!

**Review by aleos89 — APPROVED** [review]

**Comment by yozzyurubuza** [comment]
> Focused Arrow Strike no longer considers cards that increase/decrease critical damage (pre-re)

May I know the reason or reference for this change?

**Comment by Akkarinage** [comment]
> May I know the reason or reference for this change?

Official server tests were the reference for these changes.

**Pull request merged**